File: python/src/htomd/_metadata.py
"""Read explicit metadata before cleanup, then refine it from selected content."""

from __future__ import annotations

import json
from collections.abc import Iterator
from dataclasses import replace

from ._models import Metadata
from ._tree import Node, text_content, walk
from ._urls import safe_url

ARTICLE_TYPES = frozenset(
    {
        "Article",
        "NewsArticle",
        "BlogPosting",
        "TechArticle",
        "ScholarlyArticle",
        "MedicalScholarlyArticle",
        "Report",
        "AnalysisNewsArticle",
        "OpinionNewsArticle",
        "ReviewNewsArticle",
        "BackgroundNewsArticle",
        "APIReference",
        "LiveBlogPosting",
    }
)


def string(value: object) -> str | None:
    return value.strip() or None if isinstance(value, str) else None


def json_articles(value: object) -> Iterator[dict[str, object]]:
    stack = [value]
    while stack:
        item = stack.pop()
        if isinstance(item, list):
            stack.extend(reversed(item))
        elif isinstance(item, dict):
            kind = item.get("@type", [])
            kinds = [kind] if isinstance(kind, str) else kind
            if isinstance(kinds, list) and any(
                isinstance(entry, str) and entry.rsplit("/", 1)[-1] in ARTICLE_TYPES
                for entry in kinds
            ):
                yield item
            graph = item.get("@graph")
            if isinstance(graph, (dict, list)):
                stack.append(graph)


def author_name(value: object) -> str | None:
    if isinstance(value, str):
        return string(value)
    if isinstance(value, dict):
        return string(value.get("name"))
    if isinstance(value, list):
        names = [name for entry in value if (name := author_name(entry))]
        return ", ".join(names) or None
    return None


def read_jsonld(node: Node) -> dict[str, object]:
    if node.attrs.get("type", "").lower() != "application/ld+json":
        return {}
    try:
        value = json.loads(text_content(node, normalize=False))
    except (ValueError, RecursionError):
        return {}
    return next(json_articles(value), {})


def read_metadata(root: Node, url: str | None, base: str | None) -> Metadata:
    fields: dict[str, str] = {}
    article: dict[str, object] = {}
    title = language = canonical = None
    for node in walk(root):
        if node.tag == "meta":
            key = node.attrs.get("property", node.attrs.get("name", "")).lower()
            content = string(node.attrs.get("content"))
            if content:
                fields.setdefault(key, content)
        elif node.tag == "title" and title is None:
            title = string(text_content(node))
        elif node.tag == "html":
            language = string(node.attrs.get("lang") or node.attrs.get("xml:lang"))
        elif node.tag == "link" and "canonical" in node.attrs.get("rel", "").lower().split():
            canonical = canonical or safe_url(node.attrs.get("href", ""), base) or None
        elif node.tag == "script" and not article:
            article = read_jsonld(node)
    return Metadata(
        title=fields.get("og:title") or title or string(article.get("headline")),
        author=fields.get("author") or author_name(article.get("author")),
        description=fields.get("description")
        or fields.get("og:description")
        or string(article.get("description")),
        language=language or string(article.get("inLanguage")) or fields.get("og:locale"),
        published_time=fields.get("article:published_time")
        or fields.get("date")
        or string(article.get("datePublished")),
        url=url,
        canonical_url=canonical,
    )


def refine_metadata(metadata: Metadata, selected: list[Node]) -> Metadata:
    heading = None
    author = metadata.author or None
    published = metadata.published_time or None
    for root in selected:
        for node in walk(root):
            if node.tag == "h1" and heading is None:
                heading = string(text_content(node))
            if author is None and local_author(node):
                author = string(text_content(node))
            if (
                published is None
                and node.tag == "time"
                and node.attrs.get("itemprop") != "dateModified"
            ):
                published = string(node.attrs.get("datetime")) or string(text_content(node))
    return replace(
        metadata,
        title=heading or metadata.title,
        author=metadata.author or author,
        published_time=metadata.published_time or published,
    )


def local_author(node: Node) -> bool:
    if node.attrs.get("itemprop") == "author" or "author" in node.attrs.get("rel", "").split():
        return True
    tokens = node.attrs.get("class", "").lower().split()
    return any(token in {"byline", "author", "p-author"} for token in tokens)

File: go/metadata.go
package htomd

import (
	"encoding/json"
	"io"
	"strings"
)

var articleTypes = set("Article NewsArticle BlogPosting TechArticle ScholarlyArticle MedicalScholarlyArticle Report AnalysisNewsArticle OpinionNewsArticle ReviewNewsArticle BackgroundNewsArticle APIReference LiveBlogPosting")

func optional(s string) *string {
	s = strings.TrimSpace(s)
	if s == "" {
		return nil
	}
	return &s
}
func first(values ...*string) *string {
	for _, v := range values {
		if v != nil && *v != "" {
			return v
		}
	}
	return nil
}
func jsonString(v any) *string {
	s, ok := v.(string)
	if !ok {
		return nil
	}
	return optional(s)
}
func authorName(v any) *string {
	switch a := v.(type) {
	case string:
		return optional(a)
	case map[string]any:
		return jsonString(a["name"])
	case []any:
		names := []string{}
		for _, entry := range a {
			if n := authorName(entry); n != nil {
				names = append(names, *n)
			}
		}
		return optional(strings.Join(names, ", "))
	}
	return nil
}
func jsonArticle(n *node) map[string]any {
	if strings.ToLower(n.attrs["type"]) != "application/ld+json" {
		return nil
	}
	var value any
	decoder := json.NewDecoder(strings.NewReader(textContent(n, false)))
	decoder.UseNumber()
	if decoder.Decode(&value) != nil {
		return nil
	}
	var trailing any
	if decoder.Decode(&trailing) != io.EOF {
		return nil
	}
	stack := []any{value}
	for len(stack) > 0 {
		item := stack[len(stack)-1]
		stack = stack[:len(stack)-1]
		switch v := item.(type) {
		case []any:
			for i := len(v) - 1; i >= 0; i-- {
				stack = append(stack, v[i])
			}
		case map[string]any:
			kinds, ok := v["@type"].([]any)
			if !ok {
				kinds = []any{v["@type"]}
			}
			for _, kind := range kinds {
				if s, ok := kind.(string); ok {
					parts := strings.Split(s, "/")
					if articleTypes[parts[len(parts)-1]] {
						return v
					}
				}
			}
			switch graph := v["@graph"].(type) {
			case []any:
				stack = append(stack, graph)
			case map[string]any:
				stack = append(stack, graph)
			}
		}
	}
	return nil
}
func readMetadata(root *node, source, base *string) Metadata {
	fields := map[string]*string{}
	var article map[string]any
	var title, language, canonical *string
	for _, n := range walk(root) {
		switch n.tag {
		case "meta":
			key, ok := n.attrs["property"]
			if !ok {
				key = n.attrs["name"]
			}
			key = strings.ToLower(key)
			if content := optional(n.attrs["content"]); content != nil && fields[key] == nil {
				fields[key] = content
			}
		case "title":
			if title == nil {
				title = optional(textContent(n, true))
			}
		case "html":
			language = first(optional(n.attrs["lang"]), optional(n.attrs["xml:lang"]))
		case "link":
			if set(strings.ToLower(n.attrs["rel"]))["canonical"] {
				canonical = first(canonical, safeURL(n.attrs["href"], base))
			}
		case "script":
			if len(article) == 0 {
				article = jsonArticle(n)
			}
		}
	}
	var sourceCopy *string
	if source != nil {
		s := *source
		sourceCopy = &s
	}
	return Metadata{first(fields["og:title"], title, jsonString(article["headline"])), first(fields["author"], authorName(article["author"])), first(fields["description"], fields["og:description"], jsonString(article["description"])), first(language, jsonString(article["inLanguage"]), fields["og:locale"]), first(fields["article:published_time"], fields["date"], jsonString(article["datePublished"])), sourceCopy, canonical}
}
func refineMetadata(m *Metadata, selected []*node) {
	var heading *string
	author, published := m.Author, m.PublishedTime
	for _, root := range selected {
		for _, n := range walk(root) {
			if n.tag == "h1" && heading == nil {
				heading = optional(textContent(n, true))
			}
			tokens := set(strings.ToLower(n.attrs["class"]))
			if author == nil && (n.attrs["itemprop"] == "author" || set(n.attrs["rel"])["author"] || tokens["byline"] || tokens["author"] || tokens["p-author"]) {
				author = optional(textContent(n, true))
			}
			if published == nil && n.tag == "time" && n.attrs["itemprop"] != "dateModified" {
				published = first(optional(n.attrs["datetime"]), optional(textContent(n, true)))
			}
		}
	}
	m.Title = first(heading, m.Title)
	m.Author = first(m.Author, author)
	m.PublishedTime = first(m.PublishedTime, published)
}
